runtime/src/main/scala/com/armanbilge/vilcacora/runtime/Interpreter.scala

```diff
@@ -0,0 +1,305 @@
+      case other =>
+        IO.raiseError(
+          new NotImplementedError(s"Operation not implemented: ${other.getClass.getSimpleName}"),
+        )
+    }
+
+  // --- Operation Handlers ---
+
+  private def handleElementWise(op: Operation, memory: MemoryMap, model: ModelIR)(
+      f: (Float, Float) => Float,
+  ): IO[Unit] = IO {
+    val outputAllocation = model.allocations(op.outputs.head)
+    val count = outputAllocation.shape.product
+
+    val inputA = memory(op.inputs(0)).asInstanceOf[Ptr[CFloat]]
+    val inputB = memory(op.inputs(1)).asInstanceOf[Ptr[CFloat]]
+    val output = memory(op.outputs(0)).asInstanceOf[Ptr[CFloat]]
+
+    for (i <- 0 until count)
+      !(output + i) = f(!(inputA + i), !(inputB + i))
+  }
+
+  private def handleCast(op: Operation.Cast, memory: MemoryMap, model: ModelIR): IO[Unit] = IO {
+    val inputAlloc = model.allocations(op.input)
+    val outputAlloc = model.allocations(op.output)
+
+    val inputPtr = memory(op.input)
+    val outputPtr = memory(op.output)
+    val count = inputAlloc.shape.product
+
+    (inputAlloc.dataType, outputAlloc.dataType) match {
+      case (from, to) if from == to =>
+        val _ = memcpy(outputPtr, inputPtr, (count * from.sizeInBytes).toUSize)
+      case (DataType.Float64, DataType.Float32) =>
+        val in = inputPtr.asInstanceOf[Ptr[CDouble]]
+        val out = outputPtr.asInstanceOf[Ptr[CFloat]]
+        for (i <- 0 until count)
+          !(out + i) = (!(in + i)).toFloat
+      case (DataType.Float32, DataType.Float64) =>
+        val in = inputPtr.asInstanceOf[Ptr[CFloat]]
+        val out = outputPtr.asInstanceOf[Ptr[CDouble]]
+        for (i <- 0 until count)
+          !(out + i) = (!(in + i)).toDouble
+      // Add more cases for other data types
+      case (from, to) =>
+        throw new NotImplementedError(s"Cast from $from to $to is not implemented.")
+    }
+  }
+
+  private def handleSvmClassifier(
+      op: Operation.SVMClassifier,
+      memory: MemoryMap,
+      model: ModelIR,
+  ): IO[Unit] = {
+    // The number of features is the size of the last dimension of the input tensor.
+    val numFeatures = model.allocations(op.input).shape.last
+
+    val svmResources = for {
+      modelPtr <- buildSvmModelResource(op, numFeatures)
+      // Allocate space for features + 1 for the terminator node.
+      svmInputNode <- malloc[svm_node](sizeof[svm_node] * (numFeatures + 1).toUSize)
+    } yield (modelPtr, svmInputNode)
+
+    svmResources.use { case (modelPtr, svmInputNode) =>
+      // Using a for-comprehension here breaks the logic into clean, sequential IO steps.
+      for {
+        inputPtr <- IO.pure(memory(op.input).asInstanceOf[Ptr[CDouble]])
+
+        // This entire block is now a single IO action that performs the C interop.
+        predictionResult <- IO {
+          // 1. Fill the svm_node struct with input data from the input tensor.
+          for (i <- 0 until numFeatures) {
```

Comment:
We should use a `while` loop with manually-incremented counter instead of a `for` (which desugars to a `Range` allocation, `foreach`, and lambda with dynamic dispatch).